### PR TITLE
1 refactor serializer specs

### LINKED ISSUE #29: Learn Rails 13. Refactor 4.
## Overview of this refactor series.

In this series of refactor issues, we'll be exploring writing code in a consistent manner.

That consistent manner, that style, is what I use at work. As that's what I know. We'll follow it in our work together but, take/leave parts you wish to in your own work :slightly_smiling_face:  As, consistency is more important that what style it actually is.

Some of this we'll pair on, some you'll do solo.

### Linters

Code style. Easily a topic you could **waste your entire life having opinions** about. 

The benefit of code style is having consistency between authors, so it's easier to read others code.

So the only thing that matters is consistent style, not what style. In my opinion.

### Overview of these refactor issues.

_Notice how I've broken this down into steps, steps that are very small. Always worth doing._

1. Rubocop using default community configuration & passing locally.
2. Rubocop setup to run in your editor.
3. Rubocop running on each PR you push to GitHub.
4. Simple class pattern to always follow.
5. More to follow.

### Linters. Rubocop!

We've Rubocop setup in our project, lovely.

Rubocop does much heavy lifting for us. It enforces style and is highly configurable.

## Refactor 4. A simple class pattern.

This is a simple class pattern used for any Plain Old Ruby Object (PORO). It's at the bottom of this description.

Which we typically keep in `app/services` in Rails. But `lib` would be fine too.

I want you to refactor `BookSerializer` to use this pattern. I will list out the steps here, I want a PR demonstrating each step is complete please! One PR per step.

If you get stuck on any step, share what you have and I'll suggest tips :)

**Step 1. Refactor serializer specs**

Specs shouldn't use plain old Ruby methods in them. This one does currently, to build the expected response to test against.

Instead be declarative, explicitly describe the class returned by the Serializer. The hash.

https://timurnugmanov.com/posts/how-to-match-hashes-in-rspec-like-a-pro/

<details>
  <summary>Help</summary>
  
```ruby
let(:expected_response) do
  {
    # write out the hash you expect back
    # using matchers to avoid having to know the specific book attributes, like that article shows how
  }
end

it { is_expected.to(match(expected_response)) }
```
</details>

Also, read every bit of code in this spec. Do you need it?

Are you happy this spec is testing the behaviour you care about? Can it now act as validation as you TDD a refactor of the class? If not, add anything else you think you need.

<details>
  <summary>Help</summary>
It's not a model, remove that.

It does not test what happens if it's passed no books.

Rubocop compliance.

</details>

**Step 2. Refactor the Serializer**

The sample class is now ready to be TDD'd!

You have you tests, you can make lots of small incremental changes. Then validate that the tests still pass. Until you reach the desired outcome.

---

```ruby
# Adapted from a real one, used to updates Users. 
# Its called by a Controller and updates emails etc.

# app/services/users/update_user.rb
module Users
  class UpdateUser
    # All this setup above 'private' allows us to call this class like so
    # Users::UpdateUser.call(user)
    # Which looks nice than Users::UpdateUser.new(user).call
    def self.call(user)
      new(user).call
    end

    def initialize(user)
      @user = user
    end

    def call
      do_something
      do_a_complex_thing
      result
    end

    private

    attr_reader :user

    # Made up work to demo how to treat variables
    def do_something
      result = 1000 / 5
    end

    # This PORO only has a single responsibility.
    # So it calls another PORO to do something else that's complex.
    def do_a_complex_thing
      SomeOtherClass.complex_thing(result)
    end

    # A Reader
    def result
      @result ||= 0
    end
  end
end


# spec/services/users/update_user_spec.rb
RSpec.describe Users::UpdateUser do
  # We only have one public method to test, nice.
  describe '.call' do
    subject { described_class.call(user) }
    let(:user) { {} }

    it 'returns something' do
    end

    context 'SomeOtherClass raises error' do
      # etc
    end
  end
end
```

### PR DESCRIPTION
**Pull request by awalsh3**
- Adds a rubocop todo file to project as it wasn't saving and autocorrecting automatically without one!
- Also refactors to remove the plain old Ruby methods
- Adds matchers to avoid having to know the specific book attributes

Let me know if I am on the right path? Thanks!

### CODE REVIEW

File: spec/services/book_serializer_spec.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
```
Comment (patrickquigley102) [inline spec/services/book_serializer_spec.rb:2]: I love the way you've broken down your changes through the commit history :muscle:

**Comment by patrickquigley102** [comment]
Oh crap I reviewed too soon, sorry pal. I'll cancel the previous one and wait. The chances of that :joy:

**Comment by awalsh3** [comment]
Nope, my fault! I created the PR and then added more commits. I am pausing for today, now!

**Comment by patrickquigley102** [comment]
What happens if you pass a single book to this class? Not an array of one book but, just one book.

**Comment by awalsh3** [comment]
This is what happens, you guessed correctly:  

1) BookSerializer when there is only 1 book returns 1 book
     Failure/Error: return unless books.any?
     
     NoMethodError:
       undefined method `any?' for #<Book id: 1, author_name: "Default Author", title: "Default Title", description: "Default Description", created_at: "2023-11-03 16:47:41.507328000 +0000", updated_at: "2023-11-03 16:47:41.507328000 +0000", rating: 5, word_count: 500>

File: spec/services/book_serializer_spec.rb [hunk 2]
```diff
@@ -2,50 +2,70 @@
...
+        a_hash_including(
+          status: be_an_instance_of(String),
+          message: be_an_instance_of(String)
+        )
+    }
   end
 
   context 'serialization of book' do
     let(:books) { [book1, book2] }
+    subject(:serialized_result) { BookSerializer.serialize(books) }
```
Comment (patrickquigley102) [inline spec/services/book_serializer_spec.rb]: Subject is repeated identically 3 times in this spec. You can define it once and share across all the contexts.

File: spec/services/book_serializer_spec.rb [hunk 3]
```diff
@@ -2,50 +2,70 @@
...
 
-def book_attributes(book)
-  {
-    author_name: book.author_name,
-    title: book.title,
-    description: book.description,
-    rating: book.rating,
-    word_count: book.word_count
-  }
+    it 'returns 1 book' do
```
Comment (patrickquigley102) [inline spec/services/book_serializer_spec.rb]: This is not the expected result right?

This test is invalid it does not pass.

I understand you wanted to write a test for what the code should do.

But right now, I want you to write tests for what the code actually does. It raises a NoMethodError exception.

Can you write this test to assert what the behavior is please?

Then in the next MR we'll change this test to what it is here, alongside the change in the class to support the behavior we want.

**Review by patrickquigley102 — COMMENTED** [review]
2 bits of feedback pal

**Review by patrickquigley102 — APPROVED** [review]
Niiiiiiiiiiiiiice! Get it merged in mate!

The test for the NoMethodError we added in is worth visiting again.

It can feel a bit redundant, knowing that you'll be changing the code on the next PR. And as you get more familiar with these concepts in your day-to-day you'll naturally cut out steps like this, which is totally cool.

But for this exercise, and any problem which is new and unfamiliar, it's good to slow down. Break work down. Stay firm on those steps, and simplify a hard thing into many small simple things!

Go slow, to go fast.

File: spec/services/book_serializer_spec.rb (hunk 3 above)
Comment (awalsh3) [reply spec/services/book_serializer_spec.rb]: Could we go over this today? (Reminder comment)

**Pull request merged**